lib/Runtime/Library/JavascriptPromise.cpp

```diff
@@ -1462,22 +1462,33 @@ namespace Js
         }
     }
 
-    JavascriptPromise* JavascriptPromise::InitializePromise_TTD(ScriptContext* scriptContext, uint32 status, Var result, JsUtil::List<Js::JavascriptPromiseReaction*, HeapAllocator>& resolveReactions, JsUtil::List<Js::JavascriptPromiseReaction*, HeapAllocator>& rejectReactions)
+    JavascriptPromise* JavascriptPromise::InitializePromise_TTD(ScriptContext* scriptContext, uint32 status, bool isHandled, Var result, SList<Js::JavascriptPromiseReaction*, HeapAllocator>& resolveReactions,SList<Js::JavascriptPromiseReaction*, HeapAllocator>& rejectReactions)
     {
         Recycler* recycler = scriptContext->GetRecycler();
         JavascriptLibrary* library = scriptContext->GetLibrary();
 
         JavascriptPromise* promise = library->CreatePromise();
 
-        promise->status = (PromiseStatus)status;
+        promise->SetStatus((PromiseStatus)status);
+        if (isHandled)
+        {
+            promise->SetIsHandled();
+        }
         promise->result = result;
 
-        promise->resolveReactions = RecyclerNew(recycler, JavascriptPromiseReactionList, recycler);
-        promise->resolveReactions->Copy(&resolveReactions);
-
-        promise->rejectReactions = RecyclerNew(recycler, JavascriptPromiseReactionList, recycler);
-        promise->rejectReactions->Copy(&rejectReactions);
+        promise->reactions = RecyclerNew(recycler, JavascriptPromiseReactionList, recycler);
+        SList<Js::JavascriptPromiseReaction*, HeapAllocator>::Iterator resolveIterator = resolveReactions.GetIterator();
+        SList<Js::JavascriptPromiseReaction*, HeapAllocator>::Iterator rejectIterator = rejectReactions.GetIterator();
 
+        while (resolveIterator.Next() && rejectIterator.Next())
```

Comment:
Perhaps it is worth asserting that these iterators exhaust at the same time?

Reply:
fixed